File: Graphs/MinimumSpanningTree/Python/Prim'sAdjacencyList.py
import heapq

class PrimAdjacencyList:
    def __init__(self, vertices):
        self.V = vertices
        self.adj = [[] for _ in range(vertices)]

    def add_edge(self, u, v, w):
        self.adj[u].append((v, w))
        self.adj[v].append((u, w))

    def mst(self):
        key = [float('inf')] * self.V
        parent = [-1] * self.V
        in_mst = [False] * self.V
        key[0] = 0

        pq = [(0, 0)]  # (weight, vertex)

        while pq:
            _, u = heapq.heappop(pq)
            if in_mst[u]:
                continue
            in_mst[u] = True

            for v, w in self.adj[u]:
                if not in_mst[v] and w < key[v]:
                    key[v] = w
                    parent[v] = u
                    heapq.heappush(pq, (key[v], v))

        return [(parent[v], v, key[v]) for v in range(1, self.V)]

File: Graphs/MinimumSpanningTree/CPP/Prim'sAdjacencyList.cpp
#include <vector>
#include <queue>
#include <tuple>
#include <climits>
using namespace std;

class PrimAdjacencyList {
    int V;
    vector<vector<pair<int,int>>> adj;

public:
    PrimAdjacencyList(int vertices) : V(vertices) {
        adj.resize(V);
    }

    void addEdge(int u, int v, int w) {
        adj[u].push_back({v, w});
        adj[v].push_back({u, w});
    }

    vector<tuple<int,int,int>> mst() {
        vector<int> key(V, INT_MAX), parent(V, -1);
        vector<bool> inMST(V, false);
        key[0] = 0;

        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> pq;
        pq.push({0, 0});

        while (!pq.empty()) {
            int u = pq.top().second;
            pq.pop();
            if (inMST[u]) continue;
            inMST[u] = true;

            for (auto &[v, w] : adj[u]) {
                if (!inMST[v] && w < key[v]) {
                    key[v] = w;
                    parent[v] = u;
                    pq.push({key[v], v});
                }
            }
        }

        vector<tuple<int,int,int>> mstEdges;
        for (int v = 1; v < V; v++)
            mstEdges.push_back({parent[v], v, key[v]});
        return mstEdges;
    }
};
